### PR TITLE
feature/1911 speedup best trial

### PR DESCRIPTION
**Pull request by chris-chris**
study.best_trial
is soooo slow...

I just optimized the query.

### CODE REVIEW

File: optuna/storages/base.py [hunk 1]
```diff
@@ -177,18 +177,15 @@ def get_n_trials(self, study_id, state=None):
...
-
-        if len(all_trials) == 0:
-            raise ValueError('No trials are completed yet.')
+    def get_best_trial(self, study_id):
+        # type: (int) -> structs.FrozenTrial
 
-        if self.get_study_direction(study_id) == structs.StudyDirection.MAXIMIZE:
-            return max(all_trials, key=lambda t: t.value)
-        return min(all_trials, key=lambda t: t.value)
+        return self.get_best_value_trial(study_id)
```
[the remark this answers is not in the document]
Comment (chris-chris) [reply optuna/storages/base.py]: Fixed as you mentioned

File: optuna/storages/base.py [hunk 2]
```diff
@@ -177,18 +177,15 @@ def get_n_trials(self, study_id, state=None):
 
         raise NotImplementedError
 
-    def get_best_trial(self, study_id):
+    @abc.abstractmethod
+    def get_best_value_trial(self, study_id):
         # type: (int) -> structs.FrozenTrial
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (chris-chris) [reply optuna/storages/base.py]: okay

File: optuna/storages/base.py [hunk 3]
```diff
@@ -177,18 +177,17 @@ def get_n_trials(self, study_id, state=None):
 
         raise NotImplementedError
 
-    def get_best_trial(self, study_id):
+    @abc.abstractmethod
+    def get_best_value_trial(self, study_id):
```
Comment (hvy) [inline optuna/storages/base.py]: I guess we don't really need this method then 🙂

**Comment by hvy** [comment]
@chris-chris just in case you're not already aware, you can check the CI results from the job list at the bottom of this thread. It's currently complaining on e.g. unused imports which should be quite straight forward to fix.

**Comment by chris-chris** [comment]
> @chris-chris just in case you're not already aware, you can check the CI results from the job list at the bottom of this thread. It's currently complaining on e.g. unused imports which should be quite straight forward to fix.

got it.

**Comment by sile** [comment]
This PR was taken over by #729 .

**Pull request closed without merging**